### PR TITLE
Allow targeting orphan nodes

### LINKED ISSUE #3852: Cannot `plan -target` to delete resources
Running `terraform plan -target` to target a delete of a resource results in terraform thinking that there is no diff.

Example config:
```
provider "aws" {
  region = "us-east-1"
}

resource "aws_instance" "web" {
  ami = "ami-60b6c60a"
  instance_type = "t2.micro"
}
```

Run plan/apply as normal.

Comment out the lines so you end up with:

```
provider "aws" {
  region = "us-east-1"
}

/* resource "aws_instance" "web" { */
/*   ami = "ami-60b6c60a" */
/*   instance_type = "t2.micro" */
/* } */
```

Run `terraform plan`, observe that it wants to delete the instance, run `terraform plan -target aws_instance.web`, observe that it thinks there are no changes.

E.g.:
```
ip-10-254-32-102:/tmp/terraform-test% terraform plan
Refreshing Terraform state prior to plan...

aws_instance.web: Refreshing state... (ID: i-0512c5bb)

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

- aws_instance.web


Plan: 0 to add, 0 to change, 1 to destroy.
ip-10-254-32-102:/tmp/terraform-test% terraform plan -target aws_instance.web
Refreshing Terraform state prior to plan...


No changes. Infrastructure is up-to-date. This means that Terraform
could not detect any differences between your configuration and
the real physical resources that exist. As a result, Terraform
doesn't need to do anything.
```

### PR DESCRIPTION
**Pull request by jen20**
Fixes #3852.

### CODE REVIEW

File: terraform/transform_orphan.go [hunk 1]
```diff
@@ -74,8 +67,27 @@ func (t *OrphanTransformer) Transform(g *Graph) error {
...
+					prefix := fmt.Sprintf("%s.%s.%d", t.Type, t.Name, t.Index)
+					if !strings.HasPrefix(v, prefix) {
+						shouldRemove = true
+					}
+				}
+				if !shouldRemove {
+					resourceOrphans = append(resourceOrphans, resourceOrphansExcludingTargets[i])
+				}
+			}
+		}
```
Comment (phinze) [inline terraform/transform_orphan.go]: I think going about this filter in the other direction might be a bit more straightforward. Let's see how it looks:

```go
resourceOrphans := state.Orphans(config)
if len(t.Targets) > 0 {
  var targetedOrphans []string
  for _, o := range resourceOrphans {
    targeted := false
    for _, t := range t.Targets {
      prefix := fmt.Sprintf("%s.%s.%d", t.Type, t.Name, t.Index)
      if strings.HasPrefix(o, prefix) {
        targeted = true
      }
    }
    if targeted {
      targetedOrphans = append(targetedOrphans, o)
    }
  }
  resourceOrphans = targetedOrphans
}
```

Seems sort of nice to have the targeting behavior all behind the guard and the normal case is the straight-through code path. What do you think?

**Comment by phinze** [comment]
This looks 100% correct! One question about style inline, but this is some super solid graph work here. :+1:

File: terraform/transform_orphan.go (hunk 1 above)
Comment (jen20) [inline terraform/transform_orphan.go]: Likely better, I'll update.

**Comment by phinze** [comment]
LGTM

**Pull request merged**